src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,505 @@
+from typing import Dict, Optional, Union
+
+import torch
+from tqdm import tqdm
+
+from peft.utils.incremental_pca import IncrementalPCA
+
+from .config import EvaConfig, LoraConfig
+from .layer import Linear, LoraLayer
+
+
+class SVDHook:
+    """
+    A forward hook for calculating incremental SVD/PCA on layer input activations. The hook is designed to be
+    registered to a PyTorch module using the `register_forward_hook` method.
+
+    This hook performs a step of incremental Singular Value Decomposition (SVD) on the input activations of a specified
+    layer during the forward pass of a neural network. The hook also tracks convergence of the computed components
+    using cosine similarity between the current and previous components.
+
+    Attributes:
+    -----------
+    name (str): Name of the layer to which this hook is attached. n_components (int): Number of principal
+        components to compute. sim_thresh (Union[float, torch.Tensor]): Similarity threshold for convergence.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        n_components: int,
+        sim_thresh: Union[float, torch.Tensor],
+        prepare_activations_fn: Optional[callable] = None,
+    ):
+        self.name = name
+        self.n_components = n_components
+        self.sim_thresh = sim_thresh
+        if prepare_activations_fn is None:
+            self.prepare_activations_fn = self._prepare_activations_fn_default
+        else:
+            self.prepare_activations_fn = prepare_activations_fn
+
+        if isinstance(sim_thresh, torch.Tensor) and len(sim_thresh.shape) > 0:
+            check1 = sim_thresh.size(0) == n_components or sim_thresh.size(0) == 1
+            check2 = len(sim_thresh.shape) == 1
+            if not (check1 and check2):
+                raise ValueError(
+                    "if sim_thresh is a tensor with more than 0 dimensions it must have shape (n_components,) or (1,)"
+                )
+
+        self.svd = IncrementalPCA(n_components=n_components, copy=True, lowrank=True)
+
+        self.indices = None
+        self.converged = torch.zeros((n_components,), dtype=torch.bool)
+
+    def __call__(self, model, input, output):
+        previous_components = None
+        if hasattr(self.svd, "components_"):
+            previous_components = self.svd.components_.clone().detach()
+
+        states = self.prepare_activations_fn(input).detach()
+
+        # merging all but last dimension
+        if self.indices is not None:
+            states = states[self.indices]
+
+        # check if batch sizes is more than the number of components
+        if states.size(0) < self.n_components:
+            print(f"skipping SVD for {self.name} because there are less than {self.n_components} examples")
+            return
+
+        self.svd.partial_fit(states.to(torch.float32))
+
+        if previous_components is not None:
+            components = self.svd.components_
+            if len(components.shape) == 1:
```

Comment:
Could you add a short comment when this would happen?

Reply:
done, its only the case during the first svd step